src/test/java/net/sf/jabref/model/database/BibDatabaseTest.java

```diff
@@ -197,4 +197,124 @@ public void changingEntryPostsChangeEntryEvent() {
 
         assertEquals(entry, tel.getBibEntry());
     }
+
+    @Test
+    public void correctKeyCountOne() {
+        BibEntry entry = new BibEntry();
+        entry.setCiteKey("AAA");
+        database.insertEntry(entry);
+        assertEquals(database.getNumberOfKeyOccurrences("AAA"), 1);
+    }
+
+    @Test
+    public void correctKeyCountTwo() {
+        BibEntry entry = new BibEntry();
+        entry.setCiteKey("AAA");
+        database.insertEntry(entry);
+        entry = new BibEntry();
+        entry.setCiteKey("AAA");
+        database.insertEntry(entry);
+        assertEquals(database.getNumberOfKeyOccurrences("AAA"), 2);
+    }
+
+    @Test
+    public void setCiteKeySameKeySameEntry() {
+        BibEntry entry = new BibEntry();
+        entry.setCiteKey("AAA");
+        database.insertEntry(entry);
+        assertFalse(database.setCiteKeyForEntry(entry, "AAA"));
```

Comment:
I would remove the setCiteKeyForEntry method. Half of its users does not need the return value and it definitely not clear from the name of the method that the return value gives you some information about duplicates.